ksql-engine/src/main/java/io/confluent/ksql/function/UdfCompiler.java

```diff
@@ -81,22 +123,177 @@
+      throw new KsqlException("Couldn't load UDAF template", io);
+    }
+  }
 
   UdfInvoker compile(final Method method, final ClassLoader loader) {
     try {
-      final IScriptEvaluator scriptEvaluator = createScriptEvaluator(method, loader);
+      final IScriptEvaluator scriptEvaluator = createScriptEvaluator(method,
+          loader,
+          method.getDeclaringClass().getName());
       final String code = generateCode(method);
       return (UdfInvoker) scriptEvaluator.createFastEvaluator(code,
           UdfInvoker.class, new String[]{"thiz", "args"});
     } catch (final KsqlException e) {
       throw e;
-    } catch (Exception e) {
+    } catch (final Exception e) {
       throw new KsqlException("Failed to compile udf wrapper class for "
           + method, e);
     }
   }
 
+  @SuppressWarnings("unchecked")
+  KsqlAggregateFunction compileAggregate(final Method method,
+                                         final ClassLoader loader,
+                                         final String functionName) {
+    final Pair<Type, Type> valueAndAggregateTypes
+        = getValueAndAggregateTypes(method, functionName);
+    try {
+      final String generatedClassName
+          = method.getDeclaringClass().getSimpleName() + "_" + method.getName() + "_Aggregate";
+      final String udafClass = generateUdafClass(generatedClassName, method, functionName);
+      LOGGER.debug("Generated class for functionName={}, method={} class{}\n",
+          functionName,
+          method.getName(),
+          udafClass);
+      final ClassLoader javaSourceClassLoader
+          = createJavaSourceClassLoader(loader, generatedClassName, udafClass);
+
+      final IScriptEvaluator scriptEvaluator =
+          createScriptEvaluator(method,
+              javaSourceClassLoader,
+              UDAF_PACKAGE + generatedClassName);
+      final UdfArgSupplier evaluator = (UdfArgSupplier)
+          scriptEvaluator.createFastEvaluator("return new " + generatedClassName
+                  + "(args, returnType, metrics);",
+              UdfArgSupplier.class, new String[]{"args", "returnType", "metrics"});
+      return evaluator.apply(Collections.singletonList(
+          SchemaUtil.getSchemaFromType(valueAndAggregateTypes.left)),
+          SchemaUtil.getSchemaFromType(valueAndAggregateTypes.right), metrics);
+    } catch (final Exception e) {
+      throw new KsqlException("Failed to compile KSqlAggregateFunction for method='"
+          + method.getName() + "' in class='" + method.getDeclaringClass() + "'", e);
+    }
+  }
+
+  private Pair<Type, Type> getValueAndAggregateTypes(final Method method,
+                                                       final String functionName) {
+
+    final String functionInfo = "method='" + method.getName()
+        + "', functionName='" + functionName + "' UDFClass='" + method.getDeclaringClass() + "'";
+    final String invalidClass = "class='%s'"
+        + " is not supported by UDAFs. " + functionInfo;
```

Comment:
it probably makes sense to print the unspported argument type which caused this to be invalid.

Reply:
it is in the exception. See lines 219 & 222